src/Microsoft.ComponentDetection.Detectors/go/GoComponentDetector.cs

```diff
@@ -159,36 +234,53 @@ private async Task<bool> UseGoCliToScanAsync(string location, ISingleFileCompone
         return true;
     }
 
-    private void ParseGoModFile(
+    private async Task ParseGoModFileAsync(
         ISingleFileComponentRecorder singleFileComponentRecorder,
         IComponentStream file,
         GoGraphTelemetryRecord goGraphTelemetryRecord)
     {
-        using var reader = new StreamReader(file.Stream);
-
-        var line = reader.ReadLine();
-        while (line != null && !line.StartsWith("require ("))
+        void TryRegisterDependency(string s)
         {
-            if (line.StartsWith("go "))
+            if (this.TryToCreateGoComponentFromModLine(s, out var goComponent))
             {
-                goGraphTelemetryRecord.GoModVersion = line[3..].Trim();
+                singleFileComponentRecorder.RegisterUsage(new DetectedComponent(goComponent));
+            }
+            else
+            {
+                var lineTrim = s.Trim();
+                this.Logger.LogWarning("Line could not be parsed for component [{LineTrim}]", lineTrim);
+                singleFileComponentRecorder.RegisterPackageParseFailure(lineTrim);
             }
-
-            line = reader.ReadLine();
         }
 
-        // Stopping at the first ) restrict the detection to only the require section.
-        while ((line = reader.ReadLine()) != null && !line.EndsWith(")"))
+        using var reader = new StreamReader(file.Stream);
+
+        // There can be multiple require( ) sections in go 1.17+. loop over all of them.
+        while (!reader.EndOfStream)
         {
-            if (this.TryToCreateGoComponentFromModLine(line, out var goComponent))
+            var line = await reader.ReadLineAsync();
+
+            while (line != null && !line.StartsWith("require ("))
             {
-                singleFileComponentRecorder.RegisterUsage(new DetectedComponent(goComponent));
+                if (line.StartsWith("go "))
```

Comment:
nit: what if instead of having `go ` and later `line[3..]` you declare a string constant with a descriptive name and then you can do something like `line.StartsWith(MyNewDescriptiveConst)` and in the following line you do `line[MyNewDescriptiveCons.length]`.

Reply:
There are so many different private methods for all the use cases the go detector has to support that I think it's better to follow up with a separate PR to refactor this into different static helper classes.